File: sdk-python/gatedhouse/core/roles/repository.py
"""Role repository — CRUD for role definitions with org + system scoping."""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone

from gatedhouse.core.types import RoleDefinition, StoredRole
from gatedhouse.database.connection import DatabaseConnection

logger = logging.getLogger("gatedhouse.roles.repository")


class RoleRepository:
    """PostgreSQL-backed role storage."""

    def __init__(self, db: DatabaseConnection) -> None:
        self._db = db

    async def seed_base_roles(self, org_id: str, roles: list[RoleDefinition]) -> None:
        """Seed base roles for an organization (idempotent)."""
        for role in roles:
            await self._db.execute(
                """INSERT INTO gatedhouse_roles (id, org_id, name, description, permissions, inherits, is_system)
                   VALUES ($1, $2, $3, $4, $5, $6, $7)
                   ON CONFLICT (id, org_id) DO NOTHING""",
                role.key, org_id, role.name, role.description,
                json.dumps(role.permissions), json.dumps(role.inherits or []),
                role.is_system,
            )

    async def create(self, org_id: str, role: RoleDefinition) -> StoredRole:
        """Create a new role definition."""
        row = await self._db.query_one(
            """INSERT INTO gatedhouse_roles (id, org_id, name, description, permissions, inherits, is_system)
               VALUES ($1, $2, $3, $4, $5, $6, $7)
               RETURNING *""",
            role.key, org_id, role.name, role.description,
            json.dumps(role.permissions), json.dumps(role.inherits or []),
            role.is_system,
        )
        assert row is not None
        return self._to_stored(row)

    async def find_by_id(self, org_id: str, role_id: str) -> StoredRole | None:
        """Find a role by org and ID."""
        row = await self._db.query_one(
            "SELECT * FROM gatedhouse_roles WHERE org_id = $1 AND id = $2",
            org_id, role_id,
        )
        return self._to_stored(row) if row else None

    async def resolve(self, org_id: str, role_id: str) -> StoredRole | None:
        """Resolve a role, checking both org-specific and system roles."""
        # Try org-specific first
        role = await self.find_by_id(org_id, role_id)
        if role:
            return role
        # Try system role (org_id = '__system__')
        return await self.find_by_id("__system__", role_id)

    async def list_for_org(self, org_id: str) -> list[StoredRole]:
        """List all roles for an organization."""
        rows = await self._db.query(
            "SELECT * FROM gatedhouse_roles WHERE org_id = $1 ORDER BY name",
            org_id,
        )
        return [self._to_stored(r) for r in rows]

    async def update(
        self, org_id: str, role_id: str, role: RoleDefinition
    ) -> StoredRole | None:
        """Update a role definition."""
        row = await self._db.query_one(
            """UPDATE gatedhouse_roles
               SET name = $3, description = $4, permissions = $5, inherits = $6, updated_at = NOW()
               WHERE org_id = $1 AND id = $2
               RETURNING *""",
            org_id, role_id, role.name, role.description,
            json.dumps(role.permissions), json.dumps(role.inherits or []),
        )
        return self._to_stored(row) if row else None

    async def delete(self, org_id: str, role_id: str) -> bool:
        """Delete a role definition."""
        result = await self._db.execute(
            "DELETE FROM gatedhouse_roles WHERE org_id = $1 AND id = $2 AND is_system = FALSE",
            org_id, role_id,
        )
        return "DELETE 1" in result

    async def delete_all_for_org(self, org_id: str) -> None:
        """Delete all roles for an organization."""
        await self._db.execute(
            "DELETE FROM gatedhouse_roles WHERE org_id = $1", org_id
        )

    @staticmethod
    def _to_stored(row: dict) -> StoredRole:
        permissions = row["permissions"]
        if isinstance(permissions, str):
            permissions = json.loads(permissions)
        inherits = row["inherits"]
        if isinstance(inherits, str):
            inherits = json.loads(inherits)
        return StoredRole(
            id=row["id"],
            org_id=row["org_id"],
            name=row["name"],
            description=row.get("description"),
            permissions=permissions,
            inherits=inherits,
            is_system=row["is_system"],
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )

File: sdk-typescript/src/roles/repository.ts
/**
 * Role repository — CRUD operations for role definitions.
 */

import { DatabaseConnection } from '../database/connection';
import { StoredRole, RoleDefinition } from '../types';
import { createLogger } from '../logger';

const logger = createLogger('role-repository');

interface RoleRow {
  id: string;
  org_id: string;
  name: string;
  description: string | null;
  permissions: string[];
  inherits: string[];
  is_system: boolean;
  created_at: Date;
  updated_at: Date;
}

function toStoredRole(row: RoleRow): StoredRole {
  return {
    id: row.id,
    orgId: row.org_id,
    name: row.name,
    description: row.description,
    permissions: row.permissions,
    inherits: row.inherits,
    isSystem: row.is_system,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export class RoleRepository {
  constructor(private db: DatabaseConnection) {}

  /**
   * Seed base roles for an organization.
   */
  async seedBaseRoles(orgId: string, roles: RoleDefinition[]): Promise<void> {
    for (const role of roles) {
      await this.db.execute(
        `INSERT INTO gatedhouse_roles (id, org_id, name, description, permissions, inherits, is_system)
         VALUES ($1, $2, $3, $4, $5, $6, $7)
         ON CONFLICT (org_id, id) DO NOTHING`,
        [
          role.key,
          orgId,
          role.name,
          role.description ?? null,
          role.permissions,
          role.inherits ?? [],
          role.isSystem ?? false,
        ],
      );
    }
    logger.info({ orgId, count: roles.length }, 'Base roles seeded');
  }

  /**
   * Define custom roles (service-specific).
   */
  async define(orgId: string, roles: RoleDefinition[]): Promise<void> {
    for (const role of roles) {
      await this.db.execute(
        `INSERT INTO gatedhouse_roles (id, org_id, name, description, permissions, inherits, is_system)
         VALUES ($1, $2, $3, $4, $5, $6, false)
         ON CONFLICT (org_id, id) DO UPDATE SET
           name = EXCLUDED.name,
           description = EXCLUDED.description,
           permissions = EXCLUDED.permissions,
           inherits = EXCLUDED.inherits,
           updated_at = now()`,
        [
          role.key,
          orgId,
          role.name,
          role.description ?? null,
          role.permissions,
          role.inherits ?? [],
        ],
      );
    }
    logger.info({ orgId, count: roles.length }, 'Roles defined');
  }

  /**
   * Get a single role by ID and org.
   */
  async findById(orgId: string, roleId: string): Promise<StoredRole | null> {
    const row = await this.db.queryOne<RoleRow>(
      'SELECT * FROM gatedhouse_roles WHERE org_id = $1 AND id = $2',
      [orgId, roleId],
    );
    return row ? toStoredRole(row) : null;
  }

  /**
   * Find a role across system and org-specific definitions.
   * Checks org-specific first, then falls back to __system__.
   */
  async resolve(orgId: string, roleId: string): Promise<StoredRole | null> {
    const orgRole = await this.findById(orgId, roleId);
    if (orgRole) return orgRole;
    return this.findById('__system__', roleId);
  }

  /**
   * List all roles available for an organization (org-specific + system).
   */
  async listForOrg(orgId: string): Promise<StoredRole[]> {
    const rows = await this.db.query<RoleRow>(
      `SELECT * FROM gatedhouse_roles
       WHERE org_id IN ($1, '__system__')
       ORDER BY is_system DESC, name`,
      [orgId],
    );
    return rows.map(toStoredRole);
  }

  /**
   * Create a custom role.
   */
  async create(orgId: string, role: RoleDefinition): Promise<StoredRole> {
    const rows = await this.db.query<RoleRow>(
      `INSERT INTO gatedhouse_roles (id, org_id, name, description, permissions, inherits, is_system)
       VALUES ($1, $2, $3, $4, $5, $6, false)
       RETURNING *`,
      [
        role.key,
        orgId,
        role.name,
        role.description ?? null,
        role.permissions,
        role.inherits ?? [],
      ],
    );
    return toStoredRole(rows[0]);
  }

  /**
   * Update a custom role.
   */
  async update(
    orgId: string,
    roleId: string,
    updates: Partial<Pick<RoleDefinition, 'name' | 'description' | 'permissions' | 'inherits'>>,
  ): Promise<StoredRole | null> {
    const setClauses: string[] = [];
    const params: unknown[] = [];
    let idx = 3;

    if (updates.name !== undefined) {
      setClauses.push(`name = $${idx}`);
      params.push(updates.name);
      idx++;
    }
    if (updates.description !== undefined) {
      setClauses.push(`description = $${idx}`);
      params.push(updates.description);
      idx++;
    }
    if (updates.permissions !== undefined) {
      setClauses.push(`permissions = $${idx}`);
      params.push(updates.permissions);
      idx++;
    }
    if (updates.inherits !== undefined) {
      setClauses.push(`inherits = $${idx}`);
      params.push(updates.inherits);
      idx++;
    }

    if (setClauses.length === 0) return this.findById(orgId, roleId);

    setClauses.push('updated_at = now()');

    const rows = await this.db.query<RoleRow>(
      `UPDATE gatedhouse_roles SET ${setClauses.join(', ')}
       WHERE org_id = $1 AND id = $2 AND is_system = false
       RETURNING *`,
      [orgId, roleId, ...params],
    );

    return rows.length > 0 ? toStoredRole(rows[0]) : null;
  }

  /**
   * Delete a custom role. System roles cannot be deleted.
   */
  async delete(orgId: string, roleId: string): Promise<boolean> {
    const count = await this.db.execute(
      'DELETE FROM gatedhouse_roles WHERE org_id = $1 AND id = $2 AND is_system = false',
      [orgId, roleId],
    );
    return count > 0;
  }

  /**
   * Delete all roles for an organization (used on org deletion).
   */
  async deleteAllForOrg(orgId: string): Promise<void> {
    await this.db.execute(
      'DELETE FROM gatedhouse_roles WHERE org_id = $1',
      [orgId],
    );
  }
}
